### DISCUSSION TITLE
Request advice for a plugin implementation

### DISCUSSION THREAD

**Discussion by blendsdk** [verdaccio/verdaccio#2157] [Q&A]
Hi,

I would like to achieve granular control when accessing and downloading packages.

For example, an authenticated UserA is allowed to download package-v1.0.0 but NOT package-v2.0.0
I think Verdaccio does not provide such functionality by default and I have to develop a plugin for this (or a series of plugins).

To me it is not really clear where this access control should be implemented?
Is it at storage level? package.js endpoint level, or authentication level.

Any advice is greatly appreciated.

**Comment by juanpicado** [comment]
Related https://github.com/verdaccio/verdaccio/issues/226 https://github.com/verdaccio/verdaccio/issues/810
https://github.com/verdaccio/verdaccio/issues/818

> To me it is not really clear where this access control should be implemented?

In verdaccio side, which is the responsable to delegate information to the auth plugin. If you observ in our side we extract the file name but we don't use the version.
https://github.com/verdaccio/verdaccio/blob/master/src/api/middleware.js#L104

I'd considerate this change if we don't affect current plugins.

**Comment by blendsdk** [comment]
I have done some debugging and it looks like the [allow:104](https://github.com/verdaccio/verdaccio/blob/0e252fc4d7eba11472560682ed313ac16948197f/src/api/middleware.js#L109) method is eventually calling [auth:121](https://github.com/verdaccio/verdaccio/blob/0e252fc4d7eba11472560682ed313ac16948197f/src/lib/auth.js#L121). There the [pkg:124](https://github.com/verdaccio/verdaccio/blob/0e252fc4d7eba11472560682ed313ac16948197f/src/lib/auth.js#L124) is constructed and passed to the plugins.

In order to support this, the extra information should come from `middleware.js` and augmented to `pkg`. Unfortunately my knowledge of the code-base is limited to test it extensively. But it looks like the implementation is quite possible.
